Introduce docker for test

Koisell opened the pull request:
We introduce docker which allow to test locally without special
privileges. The only missing piece is to find how to publish the docker
image.

Koisell committed:
Introduce docker for test

We introduce docker which allow to test locally without special
privileges. The only missing piece is to find how to publish the docker
image.

nicflores commented on core/src/test/resources/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+RUN cat /home/fs2-ssh/nopassword.pub /home/fs2-ssh/password.pub >> /home/fs2-ssh/.ssh/authorized_keys
+
+EXPOSE 22
+CMD ["/usr/sbin/sshd", "-D"]
```
Is there an advantage to using our own Dockerfile vs using this ssh docker hub image:
https://github.com/panubo/docker-sshd ?

Koisell replied:
The advantage will be to have more control on how we parametrized our container. To tackle specific sshd version problem or special configurations.

djspiewak replied:
Honestly that's a pretty strong advantage. I really hope we never run into that kind of thing, but my trust of upstream Apache dependencies pretty much only goes so far as I can throw their logging framework. Nothing against Mina specifically, I just… worry…

nicflores replied:
Ok, yeah, I'm convinced. So the only thing we should now do is enable the docker service in `.travis.yml`:

 ```
services:
  - docker
```

Then build the image and run it during the `before_install` or `install` sections. Something like this should do the trick:

```
- cd core/src/test/resources/docker
- docker build -t fs2-ssh/sshd:test .
- docker run -d -p 222:22 --name test-sshd fs2-ssh/sshd:test
```

nicflores replied:
Forgot to mention, I think we should put all the docker items in a docker directory: `core/src/test/resources/docker`

Koisell replied:
No need of the before install, the code can do it. Both options are available which one do you prefer ?

Koisell replied:
I prefer the code one so it work out of the box for everyone

djspiewak replied:
I agree that code should work out of the box for everyone, at least as much as possible. Docker kind of corrupts this a bit, but as close as we can get would be ideal.

nicflores replied:
@Koisell, ok, I see where the code builds and runs the container, thanks. I will go ahead and merge this PR. Thanks for help Koisell.

nicflores commented on core/src/test/scala/fs2/io/ssh/ClientSpec.scala:
```diff
@@ -40,15 +40,16 @@ import java.nio.file.Paths
 
   val Timeout = 30.seconds
 
-  val TestHost = "ec2-54-205-215-42.compute-1.amazonaws.com"
```
this is great! we can remove turn off our ec2 instance that we have running just for these test, thanks!

nicflores commented:
So, I'm totally good with this PR, my only nitpick is whether or not to use: 

https://github.com/panubo/docker-sshd

instead of managing our own Dockerfile. Any thoughts @Koisell ?

Koisell commented:
I must admit I'm always on the defensive when using alpine images... Considering we are on a open source project and that Travis CI runner is itself containerised, I think we could use it.

Koisell commented:
I tried it. I'm struggling on how to make the test wait for the container to be ready... I will have a closer look later

Koisell commented:
Another issue with https://github.com/panubo/docker-sshd is that users seem  NOT to have password

djspiewak commented:
Is there a way we can simply start the docker container, then ssh in and add a user with a password and the two key files?

Koisell committed:
Build Dockerfile before test

We now build the image before running the test. It will take some time
on the first run.

Koisell committed:
Travis CI for docker

Since we changed how we run tests, we change the build accordingly.

Koisell committed:
Move docker resources in a docker subfolder

Koisell committed:
Fix before_cache build job

If there was no items the rm command was failing. This is now fixed.
